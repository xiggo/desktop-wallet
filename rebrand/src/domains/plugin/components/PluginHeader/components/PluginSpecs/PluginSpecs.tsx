import cn from "classnames";
import React from "react";
import { useTranslation } from "react-i18next";

import { Link } from "@/app/components/Link";
import { Skeleton } from "@/app/components/Skeleton";
import { usePluginStatus } from "@/domains/plugin/hooks/use-plugin-status";
import { PluginCategories } from "@/domains/plugin/plugin.contracts";

interface Properties {
	author?: string;
	category?: PluginCategories;
	isEnabled?: boolean;
	isInstalled?: boolean;
	isLoadingSize?: boolean;
	logo?: string;
	size?: string;
	url?: string;
	version?: string;
}

interface GridColProperties {
	children: React.ReactNode;
	padding?: string;
}

interface GridItemProperties {
	children: React.ReactNode;
	label: string;
	className?: string;
}

const GridItem = ({ label, children, className }: GridItemProperties) => (
	<div className={cn("flex flex-col", className)}>
		<span className="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700">{label}</span>
		<span className="font-semibold text-theme-secondary-text">{children}</span>
	</div>
);

const GridCol = ({ children, padding }: GridColProperties) => {
	const mountClassName = () => {
		let styles = "flex";

		if (padding) {
			styles = `${styles} ${padding}`;
		}

		return styles;
	};

	return <div className={mountClassName()}>{children}</div>;
};

export const PluginSpecs = ({
	author,
	category,
	url,
	version,
	size,
	isEnabled,
	isInstalled,
	isLoadingSize,
}: Properties) => {
	const { t } = useTranslation();

	const { getPluginStatus } = usePluginStatus();

	return (
		<div className="flex justify-between space-4">
			<div className="flex space-x-6 divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800">
				<GridCol>
					<GridItem label={t("COMMON.AUTHOR")}>
						<span className="font-semibold text-theme-secondary-text truncate max-w-48">{author}</span>
					</GridItem>
				</GridCol>

				<GridCol padding="pl-6">
					<GridItem label={t("COMMON.CATEGORY")}>
						{category && t(`PLUGINS.CATEGORIES.${category.toUpperCase()}`)}
					</GridItem>
				</GridCol>

				<GridCol padding="pl-6">
					<GridItem label={t("COMMON.WEBSITE")}>
						{url ? (
							<Link data-testid="PluginSpecs__website" to={url} isExternal>
								<span>{t("COMMON.VIEW")}</span>
							</Link>
						) : (
							<span>{t("COMMON.NOT_AVAILABLE")}</span>
						)}
					</GridItem>
				</GridCol>

				<GridCol padding="pl-6">
					<GridItem label={t("COMMON.STATUS")}>
						<span>{getPluginStatus({ isEnabled, isInstalled })}</span>
					</GridItem>
				</GridCol>
			</div>

			<div className="flex space-x-6 divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800">
				<GridCol padding="pl-6">
					<GridItem label={t("COMMON.SIZE")} className="text-right">
						{isLoadingSize ? (
							<span data-testid="PluginSpecs__size-skeleton">
								<Skeleton width={60} height={20} className="mt-0.5" />
							</span>
						) : (
							<span data-testid="PluginSpecs__size">{size || t("COMMON.NOT_AVAILABLE")}</span>
						)}
					</GridItem>
				</GridCol>

				<GridCol padding="pl-6">
					<GridItem label={t("COMMON.VERSION")} className="text-right">
						{version!}
					</GridItem>
				</GridCol>
			</div>
		</div>
	);
};
