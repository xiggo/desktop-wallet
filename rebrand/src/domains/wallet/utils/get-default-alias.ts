import { Networks } from "@payvo/sdk";
import { Contracts } from "@payvo/sdk-profiles";

interface GetDefaultAliasInput {
	profile: Contracts.IProfile;
	network: Networks.Network;
}

export const getDefaultAlias = ({ profile, network }: GetDefaultAliasInput): string => {
	const makeAlias = (count: number) => `${network.displayName()} #${count}`;

	const sameCoinWallets = profile.wallets().findByCoinWithNetwork(network.coin(), network.id());

	let counter = sameCoinWallets.length;

	if (counter === 0) {
		counter = 1;
	}

	while (profile.wallets().findByAlias(makeAlias(counter))) {
		counter++;
	}

	return makeAlias(counter);
};
