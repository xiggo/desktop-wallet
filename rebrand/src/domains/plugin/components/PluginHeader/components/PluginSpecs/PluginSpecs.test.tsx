import userEvent from "@testing-library/user-event";
import { ipcRenderer } from "electron";
import React from "react";

import { PluginSpecs } from "./PluginSpecs";
import { render, screen } from "@/utils/testing-library";

describe("PluginSpecs", () => {
	it("should render properly", async () => {
		const ipcRendererMock = jest.spyOn(ipcRenderer, "send").mockImplementation();

		const { asFragment } = render(
			<PluginSpecs
				author="Payvo"
				category="utility"
				url="https://github.com/arkecosystem/explorer"
				version="1.3.8"
				size="4.2 Mb"
			/>,
		);

		userEvent.click(screen.getByTestId("PluginSpecs__website"));

		expect(ipcRendererMock).toHaveBeenLastCalledWith("open-external", "https://github.com/arkecosystem/explorer");

		await expect(screen.findByText("Payvo")).resolves.toBeVisible();
		await expect(screen.findByText("Utility")).resolves.toBeVisible();
		await expect(screen.findByText("View")).resolves.toBeVisible();
		await expect(screen.findByText("1.3.8")).resolves.toBeVisible();
		await expect(screen.findByText("4.2 Mb")).resolves.toBeVisible();

		expect(asFragment()).toMatchSnapshot();

		ipcRendererMock.mockRestore();
	});

	it("should render without url and size", async () => {
		const { asFragment } = render(<PluginSpecs author="Payvo" category="utility" version="1.3.8" />);

		await expect(screen.findByText("Payvo")).resolves.toBeVisible();
		await expect(screen.findByText("Utility")).resolves.toBeVisible();

		await expect(screen.findAllByText("N/A")).resolves.toHaveLength(2);
		expect(asFragment()).toMatchSnapshot();
	});

	it("should show loading state for size", () => {
		const { container } = render(<PluginSpecs author="Payvo" category="utility" version="1.3.8" isLoadingSize />);

		expect(screen.getByTestId("PluginSpecs__size-skeleton")).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});
});
