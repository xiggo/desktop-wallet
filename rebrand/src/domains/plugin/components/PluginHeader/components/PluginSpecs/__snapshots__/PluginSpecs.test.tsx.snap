// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluginSpecs should render properly 1`] = `
<DocumentFragment>
  <div
    class="flex justify-between space-4"
  >
    <div
      class="flex space-x-6 divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800"
    >
      <div
        class="flex"
      >
        <div
          class="flex flex-col"
        >
          <span
            class="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
          >
            Author
          </span>
          <span
            class="font-semibold text-theme-secondary-text"
          >
            <span
              class="font-semibold text-theme-secondary-text truncate max-w-48"
            >
              Payvo
            </span>
          </span>
        </div>
      </div>
      <div
        class="flex pl-6"
      >
        <div
          class="flex flex-col"
        >
          <span
            class="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
          >
            Category
          </span>
          <span
            class="font-semibold text-theme-secondary-text"
          >
            Utility
          </span>
        </div>
      </div>
      <div
        class="flex pl-6"
      >
        <div
          class="flex flex-col"
        >
          <span
            class="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
          >
            Website
          </span>
          <span
            class="font-semibold text-theme-secondary-text"
          >
            <a
              class="Link__AnchorStyled-sc-ec014t-0 eeKUfa group ring-focus"
              data-ring-focus-margin="-m-1"
              data-testid="PluginSpecs__website"
              href="#"
              rel="noopener noreferrer"
              to="https://github.com/arkecosystem/explorer"
            >
              <span
                class="Link__Content-sc-ec014t-1 fPOlTq"
              >
                <span>
                  View
                </span>
              </span>
              <div
                class="sc-bdvvtL fCrhkg flex-shrink-0 duration-200 inline-block text-sm"
                data-testid="Link__external"
                height="16"
                width="16"
              >
                <svg>
                  arrow-external.svg
                </svg>
              </div>
            </a>
          </span>
        </div>
      </div>
      <div
        class="flex pl-6"
      >
        <div
          class="flex flex-col"
        >
          <span
            class="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
          >
            Status
          </span>
          <span
            class="font-semibold text-theme-secondary-text"
          >
            <span>
              Not installed
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="flex space-x-6 divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800"
    >
      <div
        class="flex pl-6"
      >
        <div
          class="flex flex-col text-right"
        >
          <span
            class="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
          >
            Size
          </span>
          <span
            class="font-semibold text-theme-secondary-text"
          >
            <span
              data-testid="PluginSpecs__size"
            >
              4.2 Mb
            </span>
          </span>
        </div>
      </div>
      <div
        class="flex pl-6"
      >
        <div
          class="flex flex-col text-right"
        >
          <span
            class="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
          >
            Version
          </span>
          <span
            class="font-semibold text-theme-secondary-text"
          >
            1.3.8
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PluginSpecs should render without url and size 1`] = `
<DocumentFragment>
  <div
    class="flex justify-between space-4"
  >
    <div
      class="flex space-x-6 divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800"
    >
      <div
        class="flex"
      >
        <div
          class="flex flex-col"
        >
          <span
            class="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
          >
            Author
          </span>
          <span
            class="font-semibold text-theme-secondary-text"
          >
            <span
              class="font-semibold text-theme-secondary-text truncate max-w-48"
            >
              Payvo
            </span>
          </span>
        </div>
      </div>
      <div
        class="flex pl-6"
      >
        <div
          class="flex flex-col"
        >
          <span
            class="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
          >
            Category
          </span>
          <span
            class="font-semibold text-theme-secondary-text"
          >
            Utility
          </span>
        </div>
      </div>
      <div
        class="flex pl-6"
      >
        <div
          class="flex flex-col"
        >
          <span
            class="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
          >
            Website
          </span>
          <span
            class="font-semibold text-theme-secondary-text"
          >
            <span>
              N/A
            </span>
          </span>
        </div>
      </div>
      <div
        class="flex pl-6"
      >
        <div
          class="flex flex-col"
        >
          <span
            class="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
          >
            Status
          </span>
          <span
            class="font-semibold text-theme-secondary-text"
          >
            <span>
              Not installed
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="flex space-x-6 divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800"
    >
      <div
        class="flex pl-6"
      >
        <div
          class="flex flex-col text-right"
        >
          <span
            class="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
          >
            Size
          </span>
          <span
            class="font-semibold text-theme-secondary-text"
          >
            <span
              data-testid="PluginSpecs__size"
            >
              N/A
            </span>
          </span>
        </div>
      </div>
      <div
        class="flex pl-6"
      >
        <div
          class="flex flex-col text-right"
        >
          <span
            class="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
          >
            Version
          </span>
          <span
            class="font-semibold text-theme-secondary-text"
          >
            1.3.8
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PluginSpecs should show loading state for size 1`] = `
<div>
  <div
    class="flex justify-between space-4"
  >
    <div
      class="flex space-x-6 divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800"
    >
      <div
        class="flex"
      >
        <div
          class="flex flex-col"
        >
          <span
            class="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
          >
            Author
          </span>
          <span
            class="font-semibold text-theme-secondary-text"
          >
            <span
              class="font-semibold text-theme-secondary-text truncate max-w-48"
            >
              Payvo
            </span>
          </span>
        </div>
      </div>
      <div
        class="flex pl-6"
      >
        <div
          class="flex flex-col"
        >
          <span
            class="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
          >
            Category
          </span>
          <span
            class="font-semibold text-theme-secondary-text"
          >
            Utility
          </span>
        </div>
      </div>
      <div
        class="flex pl-6"
      >
        <div
          class="flex flex-col"
        >
          <span
            class="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
          >
            Website
          </span>
          <span
            class="font-semibold text-theme-secondary-text"
          >
            <span>
              N/A
            </span>
          </span>
        </div>
      </div>
      <div
        class="flex pl-6"
      >
        <div
          class="flex flex-col"
        >
          <span
            class="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
          >
            Status
          </span>
          <span
            class="font-semibold text-theme-secondary-text"
          >
            <span>
              Not installed
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="flex space-x-6 divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800"
    >
      <div
        class="flex pl-6"
      >
        <div
          class="flex flex-col text-right"
        >
          <span
            class="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
          >
            Size
          </span>
          <span
            class="font-semibold text-theme-secondary-text"
          >
            <span
              data-testid="PluginSpecs__size-skeleton"
            >
              <span>
                <span
                  class="flex items-center leading-none"
                >
                  <span
                    class="react-loading-skeleton mt-0.5 css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 60px; height: 20px;"
                  >
                    ‌
                  </span>
                  ‌
                </span>
              </span>
            </span>
          </span>
        </div>
      </div>
      <div
        class="flex pl-6"
      >
        <div
          class="flex flex-col text-right"
        >
          <span
            class="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
          >
            Version
          </span>
          <span
            class="font-semibold text-theme-secondary-text"
          >
            1.3.8
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
